Basic primitives nodes

saori-eth opened the pull request:
Currently you have to manually combine them with colliders for collision but it works pretty well.

I quickly prompted an FPS arena and obstacle course here.

The diffs include a script to test all node methods which should be deleted.

https://github.com/user-attachments/assets/7d537fa9-4572-426d-a808-f7af476bf6c3

saori-eth committed:
init

saori-eth committed:
better test script

saori-eth committed:
example test all methods

saori-eth committed:
clean

saori-eth committed:
prim docs

saori-eth commented:
Stress testing with 10k primitives with no shared materials

https://github.com/user-attachments/assets/8735ed9b-959f-440a-8a1f-65ab25421bf2

```js
console.log('Starting Prim clone stress test world...')

const GRID_SIZE = 100
const SPACING = 3
const TOTAL_OBJECTS = GRID_SIZE * GRID_SIZE
const SHAPE_KINDS = ['box', 'sphere', 'cylinder', 'cone', 'torus']
const COLORS = ['#ff0000', '#00ff00', '#0000ff', '#ffff00', '#ff00ff', '#00ffff', '#ff8800', '#88ff00', '#0088ff']

console.log(`Creating ${TOTAL_OBJECTS} animated objects in a ${GRID_SIZE}x${GRID_SIZE} grid`)

const templates = {}
SHAPE_KINDS.forEach(kind => {
  templates[kind] = app.create('prim', {
    kind: kind,
    size: kind === 'sphere' ? [0.5] : kind === 'torus' ? [0.5, 0.2] : kind === 'cylinder' || kind === 'cone' ? [0.5, 1] : [1, 1, 1],
    color: COLORS[0],
    castShadow: false,
    receiveShadow: false
  })
})

const animatedObjects = []
let objectCount = 0

for (let x = 0; x < GRID_SIZE; x++) {
  for (let z = 0; z < GRID_SIZE; z++) {
    const kind = SHAPE_KINDS[Math.floor(Math.random() * SHAPE_KINDS.length)]
    const color = COLORS[Math.floor(Math.random() * COLORS.length)]
    
    const template = templates[kind]
    const instance = template.clone()
    
    const xPos = (x - GRID_SIZE / 2) * SPACING
    const zPos = (z - GRID_SIZE / 2) * SPACING
    const yPos = 2 + Math.random() * 5
    
    instance.position.set(xPos, yPos, zPos)
    instance.color = color
    
    const randomScale = 0.5 + Math.random() * 1
    instance.scale.set(randomScale, randomScale, randomScale)
    
    animatedObjects.push({
      object: instance,
      baseY: yPos,
      rotationSpeed: {
        x: (Math.random() - 0.5) * 0.05,
        y: (Math.random() - 0.5) * 0.05,
        z: (Math.random() - 0.5) * 0.05
      },
      floatSpeed: 0.5 + Math.random() * 2,
      floatAmplitude: 0.5 + Math.random() * 1.5,
      phaseOffset: Math.random() * Math.PI * 2,
      scaleSpeed: 0.5 + Math.random() * 1.5,
      scaleAmplitude: 0.1 + Math.random() * 0.3,
      baseScale: randomScale,
      orbitRadius: Math.random() > 0.7 ? 1 + Math.random() * 2 : 0,
      orbitSpeed: Math.random() * 0.02,
      colorCycleSpeed: Math.random() > 0.5 ? 0.5 + Math.random() * 2 : 0,
      colorStartIndex: Math.floor(Math.random() * COLORS.length)
    })
    
    app.add(instance)
    objectCount++
    
    if (objectCount % 50 === 0) {
      console.log(`Created ${objectCount}/${TOTAL_OBJECTS} objects...`)
    }
  }
}

Object.values(templates).forEach(template => {
  app.remove(template)
})

console.log(`All ${TOTAL_OBJECTS} objects created!`)

let globalTime = 0

app.on('update', (dt) => {
  globalTime += dt
  
  animatedObjects.forEach((item, index) => {
    const { 
      object, 
      baseY, 
      rotationSpeed, 
      floatSpeed, 
      floatAmplitude, 
      phaseOffset,
      scaleSpeed,
      scaleAmplitude,
      baseScale,
      orbitRadius,
      orbitSpeed,
      colorCycleSpeed,
      colorStartIndex
    } = item
    
    const time = globalTime + phaseOffset
    
    object.rotation.x += rotationSpeed.x
    object.rotation.y += rotationSpeed.y
    object.rotation.z += rotationSpeed.z
    
    const floatY = baseY + Math.sin(time * floatSpeed) * floatAmplitude
    
    if (orbitRadius > 0) {
      const orbitAngle = time * orbitSpeed
      const baseX = object.position.x - Math.sin(orbitAngle - dt * orbitSpeed) * orbitRadius
      const baseZ = object.position.z - Math.cos(orbitAngle - dt * orbitSpeed) * orbitRadius
      
      object.position.x = baseX + Math.sin(orbitAngle) * orbitRadius
      object.position.z = baseZ + Math.cos(orbitAngle) * orbitRadius
    }
    
    object.position.y = floatY
    
    const scaleFactor = baseScale + Math.sin(time * scaleSpeed) * scaleAmplitude * baseScale
    object.scale.set(scaleFactor, scaleFactor, scaleFactor)
    
    if (colorCycleSpeed > 0) {
      const colorProgress = (time * colorCycleSpeed * 0.1) % 1
      const colorIndex = Math.floor((colorStartIndex + colorProgress * COLORS.length) % COLORS.length)
      object.color = COLORS[colorIndex]
    }
  })
  
  
  if (globalTime > 5) {
    const fps = 1 / dt
    if (fps < 30) {
      console.log(`Performance warning: FPS dropped to ${fps.toFixed(1)}`)
    }
  }
})

console.log('\n=== Stress Test Summary ===')
console.log(`Total objects: ${TOTAL_OBJECTS}`)
console.log('Animation types:')
console.log('- Rotation (3 axes with random speeds)')
console.log('- Floating (vertical sine wave motion)')
console.log('- Scaling (pulsing size changes)')
console.log('- Orbiting (some objects orbit their position)')
console.log('- Color cycling (some objects change colors)')
console.log('\nAll objects efficiently instanced using clone()')
console.log('Watch for FPS drops to test performance limits!')
```

ashconnell commented:
hollyyyyy.

ok yeah i think there's a smart way we can do it so that its 1 draw call for every sphere no matter what size or color it is.

will probs merge this soon

saori-eth committed:
Merge remote-tracking branch 'upstream/dev' into prims

saori-eth committed:
init

saori-eth committed:
clean examples

saori-eth committed:
remove custom material option

saori-eth committed:
100k prim stress test app

saori-eth committed:
condense instnace color logic

saori-eth committed:
remove test apps

saori-eth committed:
color prop is optional

saori-eth committed:
delete notes

saori-eth committed:
instance all prims, scale instead of size

## Changes

### 1. Prim.js Modifications
**Geometry Caching Optimization**
- Changed from caching by `kind:size` to caching by `kind` only
- All primitives now use unit-sized geometries (1x1x1)
- Size is applied via scale transformation instead of creating different geometries
- Result: Dramatically reduced geometry memory usage

**Material System Simplification**
- Now uses a single white material for all primitives
- Colors are applied via instance colors in the Stage system
- Removed per-color material caching

### 2. Stage.js Enhancements
**Instance Color Support**
- Added `supportsInstanceColor` parameter to `insertLinked()`
- Added `color` parameter for per-instance colors
- New `instanceColors` Float32Array buffer for RGB values
- Custom shader modifications to inject instance color attributes


**API Changes**:
- `insertLinked()` now accepts `supportsInstanceColor` and `color` parameters
- Added `setColor()` method to staged items for runtime color changes

### 3. Documentation Updates
**Prim.md Changes**:
- Updated default color from `null` to `#ffffff`

saori-eth commented:
ngl some of this is over my head but it seems to work

## Changes

diff: https://github.com/saori-eth/hyperfy/pull/4/files

### 1. Prim.js Modifications
**Geometry Caching Optimization**
- Changed from caching by `kind:size` to caching by `kind` only
- All primitives now use unit-sized geometries (1x1x1)
- Size is applied via scale transformation instead of creating different geometries
- Result: Dramatically reduced geometry memory usage

**Material System Simplification**
- Now uses a single white material for all primitives
- Colors are applied via instance colors in the Stage system
- Removed per-color material caching

### 2. Stage.js Enhancements
**Instance Color Support**
- Added `supportsInstanceColor` parameter to `insertLinked()`
- Added `color` parameter for per-instance colors
- New `instanceColors` Float32Array buffer for RGB values
- Custom shader modifications to inject instance color attributes


**API Changes**:
- `insertLinked()` now accepts `supportsInstanceColor` and `color` parameters
- Added `setColor()` method to staged items for runtime color changes

### 3. Documentation Updates
**Prim.md Changes**:

- Updated default color from `null` to `#ffffff`

App examples:
***100k prim stress test***
https://gist.github.com/saori-eth/26f6287153038484cb76c1d58f62cdf6
***test all methods***
https://gist.github.com/saori-eth/16fa621bfc4087f43756997f5c285449

Vid is 100k prims at max fps

https://github.com/user-attachments/assets/f850fc02-66dd-4fcb-9220-d247d00434a8

saori-eth committed:
condense stage color handling

saori-eth committed:
delete test

saori-eth committed:
gpu animated colors for primitives

saori-eth committed:
emissive, better colors

saori-eth commented:
This is all really messy rn so i will rewrite and make a new PR but, i now have 100k prims, all with colors animating through gpu animated colors, some of them are emissive, running at 165 fps

# GPU Color Animation System

## 1. Per-Instance Data

Each primitive has two unique attributes stored in GPU buffers:
- **Instance Color**: The base RGB color of each object
- **Instance Phase**: A random offset (0 to 2π) that makes each object animate uniquely

## 2. Animation in the Vertex Shader

The vertex shader converts colors from RGB to HSL (Hue, Saturation, Lightness) because HSL allows for smoother, more natural color transitions:

// Time-based animations with instance phase offset
float hueAnimation = sin(uTime * uColorSpeed + instancePhase) * uHueShift;
float waveOffset = float(gl_InstanceID) / 1000.0 * 3.14159;
float waveIntensity = sin(uTime * uWaveSpeed + waveOffset) * 0.5 + 0.5;

## 3. Color Transformations

The shader applies several animated effects:

* **Hue Shifting**: Colors rotate through the spectrum using sine waves
* **Saturation Pulsing**: Colors become more/less vibrant over time
* **Lightness Variation**: Brightness oscillates subtly
* **Wave Propagation**: Creates ripple effects across instances based on their ID

## 4. Performance Benefits

With 100,000 primitives:

* **Traditional approach**: 100,000 JavaScript color updates per frame = massive CPU bottleneck
* **GPU approach**: 0 JavaScript updates, all math happens in parallel on thousands of GPU cores

## 5. The Magic

The key insight is that instead of updating colors in JavaScript and sending them to the GPU every frame, we:

1. Send the base colors once
2. Send a single time value each frame
3. Let the GPU compute all color variations in parallel

This means whether you have 100 or 100,000 objects, the performance cost is essentially the same — just updating one time uniform per frame.


https://github.com/user-attachments/assets/e5278fc8-2250-47e5-91dc-825d07ec15f9
